newrelic/structures_newrelic_one_dashboard.go

```diff
@@ -530,6 +533,24 @@ func expandDashboardLineWidgetConfigurationThresholdInput(d *schema.ResourceData
 	return lineWidgetThresholdsRoot
 }
 
+func expandDashboardTableWidgetConfigInitialSortingInput(w map[string]interface{}) *dashboards.DashboardWidgetInitialSorting {
+	var tableWidgetInitialSorting dashboards.DashboardWidgetInitialSorting
+
+	if q, ok := w["initial_sorting"]; ok && len(q.([]interface{})) > 0 {
+		dashboardInitialSortingMap := q.([]interface{})[0].(map[string]interface{})
```

Comment:
@gmanandhar-nr since you're using a [0] here (i.e. we're interested only in the first element of the list of interfaces as there can technically be only one element), you might also want to add `MaxItems: 1` in the schema of `initial_sorting` in `resource_newrelic_one_dashboard.go`

Reply:
@gmanandhar-nr in light of a recent customer concern we received for a similar attribute in this resource (an attribute that is supposed to be a list of interfaces, with the interface comprising other attributes), I wanted to give the error handling in this block a try. To do so, I made name and direction in initial_sorting optional, and then tried to declare an empty block of initial_sorting in the configuration.

```hcl
      initial_sorting {
      }
```

This ended up throwing the following error upon `terraform apply`.

```
Stack trace from the terraform-provider-newrelic plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 82 [running]:
github.com/newrelic/terraform-provider-newrelic/v2/newrelic.expandDashboardTableWidgetConfigInitialSortingInput(...)
        /Users/XXXXXXXX/myWork/src/github.com/terraform-provider-newrelic/newrelic/structures_newrelic_one_dashboard.go:540
github.com/newrelic/terraform-provider-newrelic/v2/newrelic.expandDashboardPageInput(0x14000120101?, {0x1400045a580, 0x1, 0x10340cd58?}, {0x10374c2c0, 0x14000128bc8})
        /Users/XXXXXXXX/myWork/src/github.com/terraform-provider-newrelic/newrelic/structures_newrelic_one_dashboard.go:360 +0x1004
github.com/newrelic/terraform-provider-newrelic/v2/newrelic.expandDashboardInput(0x10374c2c0?, {0x10374c2c0, 0x14000128bc8}, {0x0, 0x0})
        /Users/XXXXXXXX/myWork/src/github.com/terraform-provider-newrelic/newrelic/structures_newrelic_one_dashboard.go:33 +0x108
github.com/newrelic/terraform-provider-newrelic/v2/newrelic.resourceNewRelicOneDashboardCreate({0x1038bd6a0, 0x14000914460}, 0x0?, {0x10373d860?, 0x140006a4e00?})
        /Users/XXXXXXXX/myWork/src/github.com/terraform-provider-newrelic/newrelic/resource_newrelic_one_dashboard.go:788 +0x11c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x14000518700, {0x1038bd5f8, 0x140002cb7a0}, 0xd?, {0x10373d860, 0x140006a4e00})
        /Users/XXXXXXXX/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:707 +0xe8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x14000518700, {0x1038bd5f8, 0x140002cb7a0}, 0x140005928f0, 0x140005c4380, {0x10373d860, 0x140006a4e00})
        /Users/XXXXXXXX/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:837 +0x850
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x140005bc030, {0x1038bd5f8?, 0x140002cb680?}, 0x14000442370)
        /Users/XXXXXXXX/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xb08
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x140005c6000, {0x1038bd5f8?, 0x140002ca8a0?}, 0x14000914070)
        /Users/XXXXXXXX/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x3c0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x10385ed80?, 0x140005c6000}, {0x1038bd5f8, 0x140002ca8a0}, 0x14000914000, 0x0)
        /Users/XXXXXXXX/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x164
google.golang.org/grpc.(*Server).processUnaryRPC(0x140002721e0, {0x1038c2678, 0x14000406d00}, 0x14000910000, 0x1400058f200, 0x103f03600, 0x0)
        /Users/XXXXXXXX/go/pkg/mod/google.golang.org/grpc@v1.57.1/server.go:1358 +0xc04
google.golang.org/grpc.(*Server).handleStream(0x140002721e0, {0x1038c2678, 0x14000406d00}, 0x14000910000, 0x0)
        /Users/XXXXXXXX/go/pkg/mod/google.golang.org/grpc@v1.57.1/server.go:1735 +0x7cc
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /Users/XXXXXXXX/go/pkg/mod/google.golang.org/grpc@v1.57.1/server.go:970 +0xb4
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 41
        /Users/XXXXXXXX/go/pkg/mod/google.golang.org/grpc@v1.57.1/server.go:981 +0x160
```

Hence, we'd need a tiny bit of error handling in this block too, to make sure the interface at [0] we're reading from the list isn't nil :)

Reply:
> @gmanandhar-nr also, are we sure `initialSorting` in the `rawConfiguration` sent to the NerdGraph request to create a dashboard can only be of the format
> 
> ```json
>             "initialSorting": {
>               "direction": "desc",
>               "name": "Duration (95%)"
>             },
> ```
> 
> and not
> 
> ```json
>             "initialSorting": [
>            {
>               "direction": "desc",
>               "name": "Duration (95%)"
>             },
>            {
>               "direction": "desc",
>               "name": "Some Other Column"
>             },
>          ]
> ```
> 
> so we can be 100% sure we only need the first element in the list here

Good thought! My first thought was how could we even sort by multiple columns?! But, later realised if we have multiple rows with same value in the "primary" column that we want to sort with, we could use secondary (and tertiary and so on), to sort the sub-groups. 

However, I don't see option to set multiple column names in UI. Also, tried to send an array with 2 column names directly in the NerdGraph request. That too doesn't work. (Although it doesn't give any error in NerdGraph API Explorer(!!), in UI it shows "Select value" for both `direction` and `name`, almost as if we didn't send any values for initial sorting in the first place).

Will confirm this too with dashboard team

Reply:
@pranav-new-relic confirm with dashboard team that we can only send following format:
             

> "initialSorting": {
>               "direction": "desc",
>               "name": "Duration (95%)"
>             },

also, added a check to confirm `initialSorting` block isn't nil